mesonbuild/compilers/compilers.py

```diff
@@ -612,7 +612,7 @@ class CompilerArgs(list):
     dedup2_suffixes = ()
     dedup2_args = ()
     # Arg prefixes and args that must be de-duped by returning 1
-    dedup1_prefixes = ('-l', '-Wl,-l')
+    dedup1_prefixes = ('-l', '-Wl,-l', '-Wl,--')
```

Comment:
This looks incorrect; it will de-duplicate linker flags that push/pop state that affects how other flags are interpreted, such as `-Wl,--start-group`, `-Wl,--as-needed`, etc.

And... unfortunately, `--export-dynamic` is also one of those flags since you can use `--no-export-dynamic` to restore the default behaviour for the rest of the libraries.

I think we need to have a separate de-dup step in `.to_native()` that collapses redundant push/pop state flags. Something like: iterate over args, mark when you see a specific push-state flag, and you can remove all duplicate push-state flags till you see the matching pop-state, and then you can remove all duplicate pop-state flags till you see a push-state, and so on.

And of course it can't be that easy because we're a build system, so we also have to remember that linker flags can be passed as: `-Wl,--as-needed,--start-group` (separated by comma), so we'd need to unwrap those into separate `-Wl` arguments first. Sounds doable with an iterator.

But then the cherry on top is that we also have to account for `--Wl,--push/pop-state` which lets you set multiple state flags at once and push/pop them, and you can't unwrap those.

... but I think that's all the corner-cases.

Reply:
Is there something simpler we can do in the mean time? Like just dedup --export-dynamic? Is --no-export-dynamic something people are actually using in the real world? I would imagine it to be extremely rare, given that export dynamic is itself quite rare.

> --no-export-dynamic to restore the default behaviour for the rest of the libraries.

Is this actually correct? Based on the manpage --export-dynamic only affects the output file and there can be only one of them. Thus a combination of --export-dynamic followed by --no-export-dynamic should be identical to not having them at all (no-export-dynamic is the default).

Reply:
Ah, you're right, I misread the manpage. I guess the chances of someone relying on multiple `--export-dynamic` / `--no-export-dynamic` flags in a single invocation gathered from different sources correctly overriding each other is pretty unlikely, although I am sure *someone* out there relies on it.

I think it's a good start to unconditionally de-dup `--export-dynamic` for this release, and then revisit this later if people complain. Just have to remember the corner-cases I listed above when we do fix this properly.

Reply:
Updated to do that.